pdns/recursordist/rec-zonetocache.cc

```diff
@@ -0,0 +1,288 @@
+#ifdef HAVE_LIBCURL
+  MiniCurl mc;
+  ComboAddress local = config.d_local;
+  std::string reply = mc.getURL(config.d_sources.at(0), nullptr, local == ComboAddress() ? nullptr : &local, config.d_timeout, false, true);
+  if (config.d_maxReceivedBytes > 0 && reply.size() > config.d_maxReceivedBytes) {
+    // We should actually detect this *during* the GET
+    throw PDNSException("Retrieved data exceeds maxReceivedBytes");
+  }
+  std::istringstream stream(reply);
+  string line;
+  while (std::getline(stream, line)) {
+    lines.push_back(line);
+  }
+#endif
+  return lines;
+}
+
+void ZoneData::ZoneToCache(const RecZoneToCache::Config& config)
+{
+  if (config.d_sources.size() > 1) {
+    d_log->info("Multiple sources not yet supported, using first");
+  }
+  d_zone = DNSName(config.d_zone);
+  d_now = time(nullptr);
+
+  // We do not do validation, it will happen on-demand if an Indeterminate record is encountered when the caches are queried
+  // First scan all records collecting info about delegations ans sigs
+  // A this moment, we ignore NSEC and NSEC3 records. It is not clear to me yet under which conditions
+  // they could be entered in into the (neg)cache.
+
+  if (config.d_method == "axfr") {
+    d_log->info("Getting zone by AXFR");
+    getByAXFR(config);
+  }
+  else {
+    vector<string> lines;
+    if (config.d_method == "url") {
+      d_log->info("Getting zone by URL");
+      lines = getURL(config);
+    }
+    else if (config.d_method == "file") {
+      d_log->info("Getting zone from file");
+      lines = getLinesFromFile(config);
+    }
+    DNSResourceRecord drr;
+    ZoneParserTNG zpt(lines, d_zone);
+    zpt.setMaxGenerateSteps(0);
+
+    while (zpt.get(drr)) {
+      DNSRecord dr(drr);
+      parseDRForCache(dr);
+    }
+  }
+
+  // Rerun, now inserting the rrsets into the cache with associated sigs
+  d_now = time(nullptr);
+  for (const auto& [key, v] : d_all) {
+    const auto& [qname, qtype] = key;
+    switch (qtype) {
+    case QType::NSEC:
+    case QType::NSEC3:
+      break;
+    case QType::RRSIG:
+      break;
+    default: {
+      vector<shared_ptr<RRSIGRecordContent>> sigsrr;
+      auto it = d_sigs.find(key);
+      if (it != d_sigs.end()) {
+        for (const auto& sig : it->second) {
+          sigsrr.push_back(sig);
+        }
+      }
+      bool auth = isRRSetAuth(qname, qtype);
+      // Same decision as updateCacheFromRecords() (we do not test for NSEC since we skip those completely)
+      if (auth || (qtype == QType::NS || qtype == QType::A || qtype == QType::AAAA || qtype == QType::DS)) {
+        g_recCache->replace(d_now, qname, qtype, v, sigsrr,
+                            std::vector<std::shared_ptr<DNSRecord>>(), auth, d_zone);
+      }
+      break;
+    }
+    }
+  }
+}
+
+// Config must be a copy, so call by value!
+void RecZoneToCache::ZoneToCache(RecZoneToCache::Config config, uint64_t configGeneration)
+{
+  setThreadName("pdns-r/ztc/" + config.d_zone);
```

Comment:
Not sure what happens if the thread name is too long?

Reply:
It will generate a warning...